5-AccessControl/2-call-api-groups/SPA/src/app/group.guard.ts

```diff
@@ -0,0 +1,84 @@
+import { ActivatedRouteSnapshot, Router, RouterStateSnapshot, UrlTree } from "@angular/router";
+import { Inject, Injectable } from "@angular/core";
+import { Location } from "@angular/common";
+import { Observable, of } from "rxjs";
+import { concatMap } from "rxjs/operators";
+
+import { MsalBroadcastService, MsalGuardConfiguration, MsalService, MSAL_GUARD_CONFIG } from "@azure/msal-angular";
+import { AccountInfo } from "@azure/msal-browser";
+
+import { BaseGuard } from "./base.guard";
+import { GraphService } from "./graph.service";
+
+type AccountWithGroupClaims = AccountInfo & {
+    idTokenClaims?: {
+        groups?: string[],
+        _claim_names?: {
+            groups: string | string[]
+        },
+        _claim_sources?: {
+            src1: {
+                endpoint: string | string[]
+            }
+        }
+    }
+}
+
+@Injectable()
+export class GroupGuard extends BaseGuard {
+
+    constructor(
+        @Inject(MSAL_GUARD_CONFIG) protected override msalGuardConfig: MsalGuardConfiguration,
+        protected override msalBroadcastService: MsalBroadcastService,
+        protected override authService: MsalService,
+        protected override location: Location,
+        protected override router: Router,
+        private graphService: GraphService
+    ) {
+        super(msalGuardConfig, msalBroadcastService, authService, location, router);
+    }
+
+    override activateHelper(state?: RouterStateSnapshot, route?: ActivatedRouteSnapshot): Observable<boolean | UrlTree> {
+        let result = super.activateHelper(state, route);
+
+        const expectedGroups: string[] = route ? route.data['expectedGroups'] : [];
+
+        return result.pipe(
+            concatMap(() => {
+                let activeAccount = this.authService.instance.getActiveAccount() as AccountWithGroupClaims;
+
+                if (!activeAccount && this.authService.instance.getAllAccounts().length > 0) {
+                    activeAccount = this.authService.instance.getAllAccounts()[0] as AccountWithGroupClaims;
+                }
+
+                // check either the ID token or the graphService to see if the user is a member of any groups
+                if (!activeAccount?.idTokenClaims?.groups && this.graphService.getUser().groupIDs.length === 0) {
+                    if (activeAccount.idTokenClaims?._claim_names && activeAccount.idTokenClaims?._claim_names.groups) {
+                        window.alert('You have too many group memberships. The application will now query Microsoft Graph to get the full list of groups that you are a member of.');
```

Comment:
>window [](http://example.com/codeflow?start=24&length=6)

Differentiate the error message between overage and absence of GroupAdmin and GroupMember memberships

Reply:
done